// vue3-snackbar/dist/vue3-snackbar.js
import { h as v, computed as c, ref as _, onMounted as x, watch as D, openBlock as g, createElementBlock as y, normalizeClass as j, unref as S, normalizeStyle as q, renderSlot as h, createElementVNode as k, createVNode as $, mergeProps as Z, createCommentVNode as w, toDisplayString as L, inject as U, nextTick as J, getCurrentScope as K, onScopeDispose as Q, getCurrentInstance as F, onUnmounted as X, createBlock as A, Teleport as Y, TransitionGroup as ee, withCtx as H, Fragment as te, renderList as I, createSlots as se, normalizeProps as oe, guardReactiveProps as ae } from "vue";
var re = "M8.27,3L3,8.27V15.73L8.27,21H15.73C17.5,19.24 21,15.73 21,15.73V8.27L15.73,3M9.1,5H14.9L19,9.1V14.9L14.9,19H9.1L5,14.9V9.1M11,15H13V17H11V15M11,7H13V13H11V7", ne = "M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16", ie = "M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z", le = "M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z", ue = "M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z";
const ce = {
  transform: "rotate(var(--r, 0deg)) scale(var(--sx, 1), var(--sy, 1))"
}, E = {
  fill: "currentColor"
}, z = {
  mdi: {
    size: 24,
    viewbox: "0 0 24 24"
  },
  "simple-icons": {
    size: 24,
    viewbox: "0 0 24 24"
  },
  default: {
    size: 0,
    viewbox: "0 0 0 0"
  }
}, N = {
  name: "icon",
  props: {
    type: {
      type: String,
      default: "mdi"
    },
    faIcon: {
      type: Object,
      default: null
    },
    path: {
      type: [String, Object, Array]
    },
    size: {
      type: [Number, String],
      default: 24
    },
    viewbox: String,
    flip: {
      type: String,
      validator: (t) => ["horizontal", "vertical", "both"].includes(t)
    },
    rotate: {
      type: [Number, String],
      default: 0
    }
  },
  setup(t) {
    if (!t.path && !t.faIcon)
      return console.warn("vue3-icon requires either a 'path' or an 'fa-icon' property"), () => v("div");
    const s = c(() => {
      var d;
      return ((d = t.faIcon) == null ? void 0 : d.prefix) || t.type;
    }), o = c(() => parseInt(t.rotate, 10)), e = c(() => z[s.value] || z.default), r = c(() => parseInt(t.size, 10) || e.value.size), i = c(() => t.faIcon ? `0 0 ${t.faIcon.icon[0]} ${t.faIcon.icon[1]}` : !1), l = c(() => i.value || t.viewbox || e.value.viewbox), f = c(() => ({
      ...ce,
      "--sx": ["both", "horizontal"].includes(t.flip) ? "-1" : "1",
      "--sy": ["both", "vertical"].includes(t.flip) ? "-1" : "1",
      "--r": isNaN(o.value) ? o.value : o.value + "deg"
    })), m = c(() => {
      var d;
      return t.faIcon ? (d = t.faIcon) == null ? void 0 : d.icon[4] : t.type === "simple-icons" && typeof t.path == "object" ? t.path.path : t.path;
    }), p = c(() => s.value === "fad" ? (console.warn("vue3-icon does not currently support Duotone FontAwesome icons"), v("path")) : Array.isArray(t.path) ? v(
      "g",
      { style: { ...E } },
      t.path.map((d) => typeof d == "string" ? v("path", { d }) : v("path", { ...d }))
    ) : v("path", { d: m.value, style: { ...E } }));
    return () => v(
      "svg",
      {
        style: f.value,
        class: ["vue3-icon"],
        width: r.value,
        height: r.value,
        viewBox: l.value
      },
      [p.value]
    );
  }
}, de = { class: "vue3-snackbar-message-wrapper" }, fe = {
  key: 0,
  class: "vue3-snackbar-message-icon"
}, me = { class: "vue3-snackbar-message-content" }, pe = {
  key: 0,
  class: "vue3-snackbar-message-badge"
}, ge = { class: "vue3-snackbar-message-title" }, ve = {
  key: 0,
  class: "vue3-snackbar-message-additional"
}, be = /* @__PURE__ */ k("div", { class: "spacer" }, null, -1), ye = { class: "vue3-snackbar-message-close" }, he = {
  __name: "Vue3SnackbarMessage",
  props: {
    borderClass: {
      type: String,
      default: ""
    },
    message: {
      type: Object,
      default: () => ({})
    },
    messageClass: {
      type: String,
      default: ""
    },
    dense: {
      type: Boolean,
      default: !1
    }
  },
  emits: ["dismiss"],
  setup(t, { emit: s }) {
    const o = s, e = t;
    let r = null, i = null, l = _(!1);
    const f = () => {
      const a = !e.message.duration && !e.message.dismissible ? 4e3 : e.message.duration;
      r = setTimeout(p, a);
    };
    x(() => {
      f();
    }), D(
      () => e.message.count,
      (a) => {
        if (a === 1)
          return !1;
        clearTimeout(r), clearTimeout(i), i = setTimeout(() => {
          l.value = !1;
        }, 1e3), l.value = !0, f();
      }
    );
    const m = () => {
      r && clearTimeout(r), p();
    }, p = () => {
      o("dismiss", e.message);
    }, d = {
      success: {
        path: ie
      },
      info: {
        path: ue
      },
      warning: {
        path: ne
      },
      error: {
        path: re
      }
    }, n = c(() => {
      const a = d[e.message.type];
      return a ? (a.type = "mdi", a) : e.message.icon && typeof e.message.icon == "object" ? e.message.icon : {
        path: "",
        type: "default"
      };
    });
    return (a, u) => (g(), y("article", {
      class: j(["vue3-snackbar-message", [
        e.message.type || "custom",
        e.messageClass,
        e.borderClass,
        {
          "has-background": e.message.background,
          "has-border": e.borderClass,
          "is-dense": e.dense,
          "shake-baby-shake": S(l)
        }
      ]]),
      style: q({
        "--message-background": e.message.background,
        "--message-text-color": e.message.textColor,
        "--message-icon-color": e.message.iconColor
      })
    }, [
      h(a.$slots, "message-inner", {
        message: e.message
      }, () => [
        k("div", de, [
          n.value ? (g(), y("div", fe, [
            h(a.$slots, "message-icon", {
              message: e.message,
              icon: n.value
            }, () => [
              $(S(N), Z(n.value, { role: "img" }), null, 16)
            ])
          ])) : w("", !0),
          k("div", me, [
            h(a.$slots, "message-badge", {
              message: e.message,
              count: e.message.count
            }, () => [
              e.message.count > 1 ? (g(), y("div", pe, L(e.message.count), 1)) : w("", !0)
            ]),
            h(a.$slots, "message-content", {
              message: e.message,
              title: e.message.title,
              text: e.message.text
            }, () => [
              k("div", ge, L(e.message.title || e.message.text), 1),
              e.message.title && e.message.text ? (g(), y("div", ve, L(e.message.text), 1)) : w("", !0)
            ])
          ]),
          be,
          k("div", ye, [
            h(a.$slots, "message-close-icon", {
              message: e.message,
              isDimissible: e.message.dismissible,
              dismiss: m
            }, () => [
              e.message.dismissible !== !1 ? (g(), y("button", {
                key: 0,
                onClick: m
              }, [
                $(S(N), {
                  type: "mdi",
                  path: S(le)
                }, null, 8, ["path"])
              ])) : w("", !0)
            ])
          ])
        ])
      ])
    ], 6));
  }
}, Se = typeof window < "u" ? HTMLElement : Object, ke = {
  /* ******************************************
   * LOCATION PROPS
   ****************************************** */
  top: {
    type: Boolean,
    default: !1
  },
  bottom: {
    type: Boolean,
    default: !1
  },
  left: {
    type: Boolean,
    default: !1
  },
  right: {
    type: Boolean,
    default: !1
  },
  /* ******************************************
   * COLOUR PROPS
   ****************************************** */
  success: {
    type: String,
    default: "#4caf50"
  },
  error: {
    type: String,
    default: "#ff5252"
  },
  warning: {
    type: String,
    default: "#fb8c00"
  },
  info: {
    type: String,
    default: "#2196f3"
  },
  messageTextColor: {
    type: String,
    default: "#fff"
  },
  messageIconColor: {
    type: String,
    default: "currentColor"
  },
  /* ******************************************
   * OTHER PROPS
   ****************************************** */
  attach: {
    type: [String, Se],
    default: "body"
  },
  border: {
    type: String,
    default: "",
    validator: (t) => ["top", "bottom", "left", "right", ""].includes(t)
  },
  backgroundOpacity: {
    type: [String, Number],
    default: 0.12,
    validator: (t) => !isNaN(parseFloat(t)) && isFinite(t)
  },
  backgroundColor: {
    type: String,
    default: "currentColor"
  },
  baseBackgroundColor: {
    type: String,
    default: "#fff"
  },
  duration: {
    type: [Number, String],
    default: null
  },
  messageClass: {
    type: String
  },
  zIndex: {
    type: Number,
    default: 1e4
  },
  dense: {
    type: Boolean,
    default: !1
  },
  reverse: {
    type: Boolean,
    default: !1
  },
  groups: {
    type: Boolean,
    default: !1
  },
  shadow: {
    type: Boolean,
    default: !1
  }
};
function Ce(t) {
  return t && t.__esModule && Object.prototype.hasOwnProperty.call(t, "default") ? t.default : t;
}
var V = { exports: {} };
function B() {
}
B.prototype = {
  on: function(t, s, o) {
    var e = this.e || (this.e = {});
    return (e[t] || (e[t] = [])).push({
      fn: s,
      ctx: o
    }), this;
  },
  once: function(t, s, o) {
    var e = this;
    function r() {
      e.off(t, r), s.apply(o, arguments);
    }
    return r._ = s, this.on(t, r, o);
  },
  emit: function(t) {
    var s = [].slice.call(arguments, 1), o = ((this.e || (this.e = {}))[t] || []).slice(), e = 0, r = o.length;
    for (e; e < r; e++)
      o[e].fn.apply(o[e].ctx, s);
    return this;
  },
  off: function(t, s) {
    var o = this.e || (this.e = {}), e = o[t], r = [];
    if (e && s)
      for (var i = 0, l = e.length; i < l; i++)
        e[i].fn !== s && e[i].fn._ !== s && r.push(e[i]);
    return r.length ? o[t] = r : delete o[t], this;
  }
};
V.exports = B;
V.exports.TinyEmitter = B;
var we = V.exports, Me = we, _e = new Me();
const M = /* @__PURE__ */ Ce(_e), C = {
  $on: (...t) => M.on(...t),
  $once: (...t) => M.once(...t),
  $off: (...t) => M.off(...t),
  $emit: (...t) => M.emit(...t)
}, b = _([]), G = Symbol();
function Ne() {
  const t = U(G);
  if (!t)
    throw new Error("No Snackbar provided!");
  return t;
}
const De = {
  install: (t, s = {}) => {
    const { disableGlobals: o = !1 } = s, e = {
      add: (r) => {
        C.$emit("add", r);
      },
      clear: () => {
        C.$emit("clear");
      }
    };
    o !== !0 && (t.config.globalProperties.$snackbar = e, typeof window < "u" && (window.$snackbar = e)), t.provide(G, e);
  }
};
function Le(t) {
  return K() ? (Q(t), !0) : !1;
}
function P(t) {
  return typeof t == "function" ? t() : S(t);
}
const W = typeof window < "u" && typeof document < "u";
typeof WorkerGlobalScope < "u" && globalThis instanceof WorkerGlobalScope;
const $e = (t) => t != null;
function xe(t) {
  return t || F();
}
function Ve(t, s = !0, o) {
  xe() ? x(t, o) : s ? t() : J(t);
}
function Be(t) {
  var s;
  const o = P(t);
  return (s = o == null ? void 0 : o.$el) != null ? s : o;
}
const Oe = W ? window : void 0, Te = W ? window.document : void 0;
function Ae() {
  const t = _(!1), s = F();
  return s && x(() => {
    t.value = !0;
  }, s), t;
}
function He(t) {
  const s = Ae();
  return c(() => (s.value, !!t()));
}
function Ie(t, s, o = {}) {
  const { window: e = Oe, ...r } = o;
  let i;
  const l = He(() => e && "MutationObserver" in e), f = () => {
    i && (i.disconnect(), i = void 0);
  }, m = c(() => {
    const a = P(t), u = (Array.isArray(a) ? a : [a]).map(Be).filter($e);
    return new Set(u);
  }), p = D(
    () => m.value,
    (a) => {
      f(), l.value && e && a.size && (i = new MutationObserver(s), a.forEach((u) => i.observe(u, r)));
    },
    { immediate: !0, flush: "post" }
  ), d = () => i == null ? void 0 : i.takeRecords(), n = () => {
    f(), p();
  };
  return Le(n), {
    isSupported: l,
    stop: n,
    takeRecords: d
  };
}
function Ee(t = {}) {
  const {
    document: s = Te,
    selector: o = "html",
    observe: e = !1,
    initialValue: r = "ltr"
  } = t;
  function i() {
    var f, m;
    return (m = (f = s == null ? void 0 : s.querySelector(o)) == null ? void 0 : f.getAttribute("dir")) != null ? m : r;
  }
  const l = _(i());
  return Ve(() => l.value = i()), e && s && Ie(
    s.querySelector(o),
    () => l.value = i(),
    { attributes: !0 }
  ), c({
    get() {
      return l.value;
    },
    set(f) {
      var m, p;
      l.value = f, s && (l.value ? (m = s.querySelector(o)) == null || m.setAttribute("dir", l.value) : (p = s.querySelector(o)) == null || p.removeAttribute("dir"));
    }
  });
}
const je = {
  __name: "Vue3Snackbar",
  props: { ...ke },
  emits: ["added", "dismissed", "removed", "cleared"],
  setup(t, { emit: s }) {
    const o = Ee(), e = t, r = s, i = c(() => ({
      "is-top": e.top,
      "is-bottom": e.top === !1 && e.bottom,
      "is-left": e.left,
      "is-right": e.left === !1 && e.right,
      "is-middle": e.top === !1 && e.bottom === !1,
      "is-centre": e.left === !1 && e.right === !1,
      "has-shadow": e.shadow,
      "is-rtl": o.value === "rtl"
    })), l = c(() => ({
      "--success-colour": e.success,
      "--error-colour": e.error,
      "--warning-colour": e.warning,
      "--info-colour": e.info,
      "--snackbar-zindex": e.zIndex,
      "--background-opacity": e.backgroundOpacity,
      "--background-color": e.backgroundColor,
      "--base-background-color": e.baseBackgroundColor,
      "--message-text-color": e.messageTextColor,
      "--message-icon-color": e.messageIconColor
    })), f = c(() => e.border ? `border-${e.border}` : ""), m = (n) => Math.abs(n.split("").reduce((a, u) => (a << 5) - a + u.charCodeAt(0) | 0, 0));
    let p = 1;
    C.$on("add", (n) => {
      r("added", n), n.group || (n.group = m(`${n.type}${n.title}${n.text}`).toString(16)), e.duration && !n.duration && n.duration !== 0 && (n.duration = +e.duration);
      const a = n.group && b.value.find((u) => u.group === n.group);
      if (e.groups === !1 || !a) {
        const u = {
          ...n,
          id: p,
          count: 1
        };
        e.reverse ? b.value.unshift(u) : b.value.push(u), p++;
      } else
        a.count++;
    }), C.$on("clear", () => {
      r("cleared"), b.value = [];
    }), X(() => {
      C.$off("add"), C.$off("clear");
    });
    const d = (n, a = !1) => {
      r(a ? "dismissed" : "removed", n), b.value = b.value.filter((u) => u.id !== n.id);
    };
    return (n, a) => (g(), A(Y, {
      to: e.attach
    }, [
      k("section", {
        id: "vue3-snackbar--container",
        class: j([[i.value], "vue3-snackbar"]),
        style: q(l.value)
      }, [
        $(ee, {
          name: "vue3-snackbar-message",
          tag: "div"
        }, {
          default: H(() => [
            (g(!0), y(te, null, I(S(b), (u) => (g(), A(he, {
              key: u.id,
              message: u,
              "message-class": e.messageClass,
              dense: e.dense,
              "border-class": f.value,
              onDismiss: a[0] || (a[0] = (O) => d(O, !0))
            }, se({ _: 2 }, [
              I(n.$slots, (O, T) => ({
                name: T,
                fn: H((R) => [
                  h(n.$slots, T, oe(ae(R)))
                ])
              }))
            ]), 1032, ["message", "message-class", "dense", "border-class"]))), 128))
          ]),
          _: 3
        })
      ], 6)
    ], 8, ["to"]));
  }
};
export {
  b as SnackbarMessages,
  De as SnackbarService,
  je as Vue3Snackbar,
  Ne as useSnackbar
};
